src/libc/stdio/fparseln.3
.\"	$NetBSD: fparseln.3,v 1.2 2003/01/18 11:29:52 thorpej Exp $
.\"
.\" Copyright (c) 1997 Christos Zoulas.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Christos Zoulas.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd November 30, 2002
.Dt FPARSELN 3
.Os
.Sh NAME
.Nm fparseln
.Nd return the next logical line from a stream
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdio.h
.Ft "char *"
.Fo "fparseln"
.Fa "FILE *stream" "size_t *len" "size_t *lineno"
.Fa "const char delim[3]" "int flags"
.Fc
.Sh DESCRIPTION
The
.Fn fparseln
function
returns a pointer to the next logical line from the stream referenced by
.Fa stream .
This string is
.Dv NUL
terminated and it is dynamically allocated on each invocation.
It is the responsibility of the caller to free the pointer.
.Pp
By default, if a character is escaped, both it and the preceding escape
character will be present in the returned string.
Various
.Fa flags
alter this behaviour.
.Pp
The meaning of the arguments is as follows:
.Bl -tag -width "lineno"
.It Fa stream
The stream to read from.
.It Fa len
If not
.Dv NULL ,
the length of the string is stored in the memory location to which it
points.
.It Fa lineno
If not
.Dv NULL ,
the value of the memory location to which is pointed to, is incremented
by the number of lines actually read from the file.
.It Fa delim
Contains the escape, continuation, and comment characters.
If a character is
.Dv NUL
then processing for that character is disabled.
If
.Dv NULL ,
all characters default to values specified below.
The contents of
.Fa delim
is as follows:
.Bl -tag -width "delim[0]"
.It Fa delim[0]
The escape character, which defaults to
.Cm \e ,
is used to remove any special meaning from the next character.
.It Fa delim[1]
The continuation character, which defaults to
.Cm \e ,
is used to indicate that the next line should be concatenated with the
current one if this character is the last character on the current line
and is not escaped.
.It Fa delim[2]
The comment character, which defaults to
.Cm # ,
if not escaped indicates the beginning of a comment that extends until the
end of the current line.
.El
.It Fa flags
If non-zero, alter the operation of
.Fn fparseln .
The various flags, which may be
.Em or Ns -ed
together, are:
.Bl -tag -width "FPARSELN_UNESCCOMM"
.It Dv FPARSELN_UNESCCOMM
Remove escape preceding an escaped comment.
.It Dv FPARSELN_UNESCCONT
Remove escape preceding an escaped continuation.
.It Dv FPARSELN_UNESCESC
Remove escape preceding an escaped escape.
.It Dv FPARSELN_UNESCREST
Remove escape preceding any other character.
.It Dv FPARSELN_UNESCALL
All of the above.
.El
.Pp
.El
.Sh RETURN VALUES
Upon successful completion a pointer to the parsed line is returned;
otherwise,
.Dv NULL
is returned.
.Pp
The
.Fn fparseln
function uses internally
.Xr fgetln 3 ,
so all error conditions that apply to
.Xr fgetln 3 ,
apply to
.Fn fparseln .
In addition
.Fn fparseln
may set
.Va errno
to
.Bq Er ENOMEM
and return
.Dv NULL
if it runs out of memory.
.Sh SEE ALSO
.Xr fgetln 3
.Sh HISTORY
The
.Fn fparseln
function first appeared in
.Nx 1.4 .


RELEASE-NOTES
Release notes for BitKeeper version 4.1 (released 20-Sep-2007)

Major features
    BAM support.  BAM stands for "Binary Asset Management" and it adds 
    support to BK for versioning large binaries.  It solves two problems:
    a) one or more binary files that are frequently changed.
    b) collections of many large binaries where you only need a subset.
    The way it solves this is to introduce the concept of BAM server[s].
    A BAM server manages a collection of binaries for one or more BAM
    clients.  BAM clients may have no data present; when it is needed
    the data is fetched from the BAM server.  
    
    In the first case above, only the tip will be fetched.  Imagine that
    you have 100 deltas, each 10MB in size.  The history is 1GB but you
    only need 10MB in your clone.

    In the second case, imagine that you have thousands of game assets
    distributed across multiple directories.  You typically work only
    in one directory at a time.  You will only need to fetch the subset
    of files that you need, the rest of the repository will have the
    history of what changed but no data (so bk log will work but 
    bk cat will have to go fetch the data).

    BAM files are currently limited to 32 bits in size.

    See bk help Howto-BAM and bk help bam for more information.

    -----

    bk repair makes use of the new -@<url> and BAM infrastructure to run
    much more quickly.  Our stress tests for this amount to a loop, cloning
    each tagged version of BitKeeper, removing each top level directory,
    and running bk repair to restore.  Works fine, and works quickly.
    Please let us know if you find problems with this feature, for now it
    is enabled when you run check with -ff (2 "f" chars) but in the future
    we intend to turn it on in any repo with autofix configured.
    See bk help repair for more information.

Summary of changes

Performance fixes
    Repositories with large numbers of changesets will run faster when
    doing pull/push/changes.  For example, in the Linux kernel 2.6 tree
    with 128 thousand changesets, the time to do "bk changes -r+" is 1/2
    of what it was in bk-4.0.2.  Null pulls/pushes should see similar
    performance gains.

    Pulls which update repositories with large numbers of deleted files
    will run significantly faster.

Windows Vista support
    The Windows version of BitKeeper now runs on Windows Vista on 32-bit
    architectures.  64-bit support for Vista and/or XP will be in a future
    release.

Triggers
    This release introduces the concept of a trigger path.  The trigger
    path may be set in the configuration file to create a set of directories
    to search.  All triggers found in all directories are run (unless one
    is a pre- trigger and it fails).
    See bk help config-etc.

    This release introduces support for pre-undo triggers.  
    See bk help triggers.

Checkout mode
    A new checkout mode, checkout:last, has been added.  For clones, it
    is the same as checkout:none, but after the file has been checked out
    (read-only or read-write), BK will maintain that state.  The main
    use is for repositories with large numbers of large BAM binaries.
    The mode makes it possible to work on a subset of those files which
    gives you the effect of a sparse tree.  If said repository was in
    checkout:get or checkout:edit mode then the tips of all files will
    be present, i.e., the repository will not be sparse.

BK/Web
    BK/Web now uses MD5 keys for all URLs and revision numbers.  This
    allows bookmarks to continue to work even after revision numbers are
    changed by merges.  For this reason, the "Link to this page" links have
    been removed.
    Note the URL syntax has not changed from the 4.0 release, so old URLs
    will still work correctly.
    
    BK/Web now lists extra files while browsing the source tree; they are
    at the end of the directory listing which is { dirs, BK files, extras }.
    Extra files are only listed if they are not ignored.  Extra files 
    may be clicked on and viewed.

    BK/Web has links on the file names in the history/annotate pages that
    will let you download a copy of that file with no markups.

SSH URLs
    BitKeeper now allows an SSH URL to specify a non-standard port,
    for example ssh://host.example.com:5555/repopath.  See bk help url.

Remote command execution
    The command syntax 'bk -@URL command' allows a bk command to be
    run on a remote repository with the output returned locally.
    See 'bk help bk' for more information.
    Example showing work in progress (diffs and new files) in a remote repo:

    	bk -@bk://work/BAM -r diffs -Nur

Configuration
    nosync: on

    By default, BK does a sync in the resolver after updating your repository.
    The reason it does not fsync() individual files is that on some Linux
    file systems, the fsync() is like a sync() and sync() is global.  Doing
    it over and over again hurt server performance.  So we do one sync after
    applying all the files.

    Some customers prefer performance over safety; they can set this variable
    and get no syncs at all.  We don't recommend that.

bk admin
    The -H option formerly checked sfile structure, limited to AT&T
    SCCS features, and checked whether the file contents were ASCII.
    As BitKeeper now fully supports non-ASCII characters, this option
    is deprecated.  Use -h, -hh, or -hhh instead.  See bk help admin.

bk bkd
    The -p option now allows you to restrict allowed connections to
    those originating from a specified IP address.  A common usage is
    "-p127.0.0.1:", which accepts connections only on the default port
    (14690) on localhost.  See bk help bkd.

bk changes
    The search facility ("-/<pattern>/") now supports two additional
    suffixes:  "g", which indicates that <pattern> is a glob, not a regular
    expression; and "t", which restricts the search to tag names.  See bk
    help changes.


src/libc/stdio/fparseln.c
/*	$NetBSD: fparseln.c,v 1.6 2007/03/07 15:12:01 drochner Exp $	*/

/*
 * Copyright (c) 1997 Christos Zoulas.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Christos Zoulas.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
__RCSID("$NetBSD: fparseln.c,v 1.6 2007/03/07 15:12:01 drochner Exp $");
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#ifdef __weak_alias
__weak_alias(fparseln,_fparseln)
#endif

#if ! HAVE_FPARSELN

#ifndef HAVE_NBTOOL_CONFIG_H
#include "reentrant.h"
#include "local.h"
#else
#define FLOCKFILE(fp)
#define FUNLOCKFILE(fp)
#endif

#if defined(_REENTRANT) && !HAVE_NBTOOL_CONFIG_H
#define __fgetln(f, l) __fgetstr(f, l, '\n')
#else
#define __fgetln(f, l) fgetln(f, l)
#endif

static int isescaped(const char *, const char *, int);

/* isescaped():
 *	Return true if the character in *p that belongs to a string
 *	that starts in *sp, is escaped by the escape character esc.
 */
static int
isescaped(const char *sp, const char *p, int esc)
{
	const char     *cp;
	size_t		ne;

	_DIAGASSERT(sp != NULL);
	_DIAGASSERT(p != NULL);

	/* No escape character */
	if (esc == '\0')
		return 0;

	/* Count the number of escape characters that precede ours */
	for (ne = 0, cp = p; --cp >= sp && *cp == esc; ne++)
		continue;

	/* Return true if odd number of escape characters */
	return (ne & 1) != 0;
}


/* fparseln():
 *	Read a line from a file parsing continuations ending in \
 *	and eliminating trailing newlines, or comments starting with
 *	the comment char.
 */
char *
fparseln(FILE *fp, size_t *size, size_t *lineno, const char str[3], int flags)
{
	static const char dstr[3] = { '\\', '\\', '#' };

	size_t	s, len;
	char   *buf;
	char   *ptr, *cp;
	int	cnt;
	char	esc, con, nl, com;

	_DIAGASSERT(fp != NULL);

	len = 0;
	buf = NULL;
	cnt = 1;

	if (str == NULL)
		str = dstr;

	esc = str[0];
	con = str[1];
	com = str[2];
	/*
	 * XXX: it would be cool to be able to specify the newline character,
	 * but unfortunately, fgetln does not let us
	 */
	nl  = '\n';

	FLOCKFILE(fp);

	while (cnt) {
		cnt = 0;

		if (lineno)
			(*lineno)++;

		if ((ptr = __fgetln(fp, &s)) == NULL)
			break;

		if (s && com) {		/* Check and eliminate comments */
			for (cp = ptr; cp < ptr + s; cp++)
				if (*cp == com && !isescaped(ptr, cp, esc)) {
					s = cp - ptr;
					cnt = s == 0 && buf == NULL;
					break;
				}
		}

		if (s && nl) { 		/* Check and eliminate newlines */
			cp = &ptr[s - 1];

			if (*cp == nl)
				s--;	/* forget newline */
		}

		if (s && con) {		/* Check and eliminate continuations */
			cp = &ptr[s - 1];

			if (*cp == con && !isescaped(ptr, cp, esc)) {
				s--;	/* forget continuation char */
				cnt = 1;
			}
		}

		if (s == 0) {
			/*
			 * nothing to add, skip realloc except in case
			 * we need a minimal buf to return an empty line
			 */
			if (cnt || buf != NULL)
				continue;
		}

		if ((cp = realloc(buf, len + s + 1)) == NULL) {
			FUNLOCKFILE(fp);
			free(buf);
			return NULL;
		}
		buf = cp;

		(void) memcpy(buf + len, ptr, s);
		len += s;
		buf[len] = '\0';
	}

	FUNLOCKFILE(fp);

	if ((flags & FPARSELN_UNESCALL) != 0 && esc && buf != NULL &&
	    strchr(buf, esc) != NULL) {
		ptr = cp = buf;
		while (cp[0] != '\0') {
			int skipesc;

			while (cp[0] != '\0' && cp[0] != esc)
				*ptr++ = *cp++;
			if (cp[0] == '\0' || cp[1] == '\0')
				break;

			skipesc = 0;
			if (cp[1] == com)
				skipesc += (flags & FPARSELN_UNESCCOMM);
			if (cp[1] == con)
				skipesc += (flags & FPARSELN_UNESCCONT);
			if (cp[1] == esc)
				skipesc += (flags & FPARSELN_UNESCESC);
			if (cp[1] != com && cp[1] != con && cp[1] != esc)
				skipesc = (flags & FPARSELN_UNESCREST);

			if (skipesc)
				cp++;
			else
				*ptr++ = *cp++;
			*ptr++ = *cp++;
		}
		*ptr = '\0';
		len = strlen(buf);
	}

	if (size)
		*size = len;
	return buf;
}

#ifdef TEST

int main(int, char **);

int
main(int argc, char **argv)
{
	char   *ptr;
	size_t	size, line;

	line = 0;
	while ((ptr = fparseln(stdin, &size, &line, NULL,
	    FPARSELN_UNESCALL)) != NULL)
		printf("line %d (%d) |%s|\n", line, size, ptr);
	return 0;
}

/*

# This is a test
line 1
line 2 \
line 3 # Comment
line 4 \# Not comment \\\\

# And a comment \
line 5 \\\
line 6

*/

#endif /* TEST */
#endif	/* ! HAVE_FPARSELN */
